### PR TITLE
profile code execution times using CUDA events

### PR DESCRIPTION
**Pull request by Simontwice**
# Description
based on https://discuss.pytorch.org/t/unexpected-execution-time-difference-for-identical-operations-on-gpu/188681/6

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review

### CODE REVIEW

File: research/conditional/utils/layer_manager.py [hunk 1]
```diff
@@ -98,7 +98,7 @@ def update_cache_for_logging(self, key, value):
             else:
                 self.logging_cache[key] = value.clone().detach().cpu()
 
-    def _combine_to_dict_key(key, layer_type, block_number):
+    def _combine_to_dict_key(self, key, layer_type, block_number):
```
Comment (sebastianjaszczur) [inline research/conditional/utils/layer_manager.py:101]: It just didn't work before, right? Insane. Unittests would be nice i guess.

File: research/conditional/utils/layer_manager.py [hunk 2]
```diff
@@ -129,20 +129,28 @@ def log_heavy(self):
...
     """
-    if obj.logging_switch and torch.cuda.is_available():
-        torch.cuda.synchronize()
-    start_time = time.time()
-
+    if layer.logging_switch:
+        if torch.cuda.is_available():
+            start = torch.cuda.Event(enable_timing=True)
+            end = torch.cuda.Event(enable_timing=True)
+            start.record()
```
Comment (sebastianjaszczur) [inline research/conditional/utils/layer_manager.py:143]: don't we need cuda.synchronize before recording? or maybe we don't need it after end.record? I'm not sure how does it work exactly.

**Review by sebastianjaszczur — APPROVED** [review]
LGTM, but see comments before merging. Thanks!

File: research/conditional/utils/layer_manager.py (hunk 2 above)
Comment (Simontwice) [reply research/conditional/utils/layer_manager.py:143]: CUDA and CPU are async, and all we need to do is send time markers to CUDA before and aftre the operation we want to measure. Afterwards, we force the CPU to wait until the CUDA operations are finished and synced to the CPU, so we get actual values of the time markers.

Read more here: https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964/11?u=szymon_antoniak

File: research/conditional/utils/layer_manager.py (hunk 1 above)
Comment (Simontwice) [reply research/conditional/utils/layer_manager.py:101]: yuh

File: research/conditional/utils/layer_manager.py (hunk 2 above)
Comment (sebastianjaszczur) [reply research/conditional/utils/layer_manager.py:143]: ok, nice

**Review by sebastianjaszczur — COMMENTED** [review]
lgtm, thank you!

**Pull request merged**